crosshair/libimpl/builtinslib.py

```diff
@@ -1676,7 +1676,12 @@ def split(self, sep:Optional[str] = None, maxsplit:int = -1):
         new_maxsplit = -1 if maxsplit == -1 else maxsplit - 1
         ret.extend(self[first_occurance+1:].split(sep=sep, maxsplit=new_maxsplit))
         return ret
-
+    
+    def zfill(self, width: int):
+        size = SmtInt(z3.Length(self.var)) # Why does this need to be casted to SmtInt?
+        offset = width - size # Unsure about the type of offset at this point
+        buffer = SmtStr("0") * offset
```

Comment:
Ah, this is a gotcha; the constructor of all Smt* types take a first argument of type `Union[str, z3.ExprRef]`.

When you pass a string, you are naming a fresh z3 variable (all z3 variables must have names). This is how we create, say, an unconstrained  integer to pass into a user function.

That distinction is super confusing here. `SmtStr("0")` is actually creating a new symbolic string with an arbitrary, unknown value and having the **name** "0".  Instead, you could do `SmtStr(z3.StringVal("0"))`.

That said, the `*` operator is overridden by SmtInt (via `__rmul__`) and already does the symbolic work you need ([check it out here, for fun](https://github.com/pschanely/CrossHair/blob/94df30a5a4dd3a7956b34909dfc2ce9f2179eb71/crosshair/libimpl/builtinslib.py#L616-L626)); you might be able to just say `buffer = "0" * offset`.

Taking all my suggestions together, you may be surprised, saying "wait, I'm just writing an implementation of zfill in regular old python - I'm not using z3 at all!" Indeed, a lot of the CrossHair codebase is just re-implementing the python library in regular old python (instead of the C-based implementation that it ships with). We just take special care to ensure we're always building on top of python that we already symbolically handle. This is a good way to start, but I'm happy to help you find something next that will require digging into the solver!

Reply:
Thanks for all the detailed explanation, to this and my other questions :) 🎉 

I'm glad to hear that so many primitive operations are already mapped to their symbolical equivalents. Indeed, it's kinda magical to implement more functionality in regular "good old python" have z3 understand how to test that so quickly.

One follow-up question about `__mul__`, found [here](https://github.com/pschanely/CrossHair/blob/94df30a5a4dd3a7956b34909dfc2ce9f2179eb71/crosshair/libimpl/builtinslib.py#L616-L626).

The `self.statespace` variable caught my eye. Is that _just_ the space / conjunction of constraints? It would appear so, but I wanted to make sure. It would also appear that the statespace is saved in thread-local storage, which I found noteworthy: https://github.com/pschanely/CrossHair/blob/94df30a5a4dd3a7956b34909dfc2ce9f2179eb71/crosshair/statespace.py#L127-L130
Any reason why?

I hope these are not too bothersome to answer. If so, feel free to just leave it at "you'll understand later" and we can move on :) . Thanks again!

Reply:
Please continue to ask questions! I'm hoping to steal from these responses to build a developer's guide :)

Maybe a preliminary that might or might not be obvious: CrossHair doesn't do any kind of AST analysis or bytecode manipulation. It just repeatedly calls the function you want to analyze and passes in special objects that are made to look like the types that the function actually takes. When stuff happens to those special objects, we consult the SMT solver and then produce other special objects, and the function continues.

The statespace is a thread local that's created for each execution of the analyzed function; it serves a few purposes that are central to how CrossHair works, but the most important two are:
(1) It stores the current set of z3 constraints for this execution. (it holds a `z3.Solver`)
(2) It consults and updates an execution branch tree. This tree represents the different execution paths that the analyzed function could take. It uses this tree to avoid taking branches that it's already explored on prior executions and updates the tree with the novel parts of its own execution.

The statepsace almost doesn't need to be a thread-local (and wasn't until recently), but there is a case or two where it's challenging to get at it. One problem is that since CrossHair actually calls the function under analysis, you may find yourself in part of the stack where you can't easily get at the statespace. (imagine that you need to invent a new symbolic value out of nowhere, for instance) Furthermore, it would always be an error to work with values from different statespaces, so we aren't really losing a lot of flexibility by making it thread-local.